### PR TITLE
Feature: Store/show masks used on specific sites in the context menu

### PR DESCRIPTION
**Pull request by maxxcrawford**
Follow-up feature for https://github.com/mozilla/fx-private-relay-add-on/pull/334

## Testing

I've outlined how to test this feature stand-alone style. However, I recommend to also mix it up with a combination of alt/right clicking AND clicking the Firefox Relay icon (in-page menu). See the previous PR for steps to test that. 

### Free account: 

- Start with an empty account (with sync enabled)
- Create two-three masks from the dashboard page
  - _Optional: Set a label_
- Go to a website with an email input form (Example: Facebook.com) 
- Alt-click (right-click) on an input field
- **Expected:** The content menu should open, with a Firefox Relay menu item (See example screenshot)
  - There should be the following menu items: Generate new mask, Recent masks, Manage all masks, Get unlimited masks
  - _Note: If you set a label, it should show the label in the "Recent masks" list, rather than the mask address._
- Click on a mask in the "Recent masks" list
- **Expected:** The mask address should be auto-filled into the email input
- Alt-click (right-click) on an input field again
- **Expected:** The content menu should open, with a Firefox Relay menu item, with a new menu entry: "Use existing mask from this site..."
- Click on "Generate new mask" button
- **Expected:** A mask address should be created and then auto-filled into the email input
- Alt-click (right-click) on an input field again
- **Expected:** There should be two items in the "Use existing mask from this site..." list 
  - _Note: The newest mask address should have a label of the current website you're on._

### Premium Account

The behavior should be the same for premium accounts, with two additional notes: 
- You can also use custom domain created masks "github@example.mozmail.com"
- There should be no "Get unlimited masks" upgrade menu item 

## Screenshots

image

image

image

### CODE REVIEW

**Review by maxxcrawford — COMMENTED** [review]
Self-review

File: src/js/background/context-menu.js [hunk 1]
```diff
@@ -86,65 +113,87 @@ const relayContextMenus = {
...
-          createExistingAliases: true,
-          parentMenu: staticMenuData.useExistingAlias,
-          exisitingSite: false,
-          currentWebsite
-        }, aliases)
+      // Create Use Existing Alias submenu
+      if (
+        currentWebsite &&
+        masksWereGeneratedOrUsedOnCurrentWebsite &&
+        userHasSomeAliasesCreated
```
Comment (lloan) [inline src/js/background/context-menu.js:139]: praise: thank you for creating human-readable variables! Makes it so much easier to understand what is going on.

File: src/js/other-websites/inpage_menu.js [hunk 2]
```diff
@@ -311,32 +323,35 @@ function hasMaskBeenUsedOnCurrentSite(mask, domain) {
...
 
 function haveMasksBeenUsedOnCurrentSite(masks, domain) {
-  return masks.some(mask => {
-    
+  return masks.some((mask) => {
     const domainList = mask.used_on;
 
     // Short circuit out if there's no used_on entry
-    if (domainList === null || domainList === "" ||  domainList === undefined) { return false; }
+    if (domainList === null || domainList === "" || domainList === undefined) {
```
Comment (lloan) [inline src/js/other-websites/inpage_menu.js:344]: suggestion (non-blocking): I've seen this logic at least a handful of times - I wonder if we should have a function that does this check and reuse it.

**Review by lloan — APPROVED** [review]
I tested this and it works.

File: src/js/background/context-menu.js [hunk 3]
```diff
@@ -153,15 +172,42 @@ const relayContextMenus = {
...
+    
+      const currentMaskType = selectedAliasObject[0].mask_type;
+      const currentUsedOnValue = selectedAliasObject[0].used_on;
+    
+      const currentPage = new URL(tab.url);
+      const currentPageHostName = currentPage.hostname;
+ 
+      // If the used_on field is blank, then just set it to the current page/hostname. Otherwise, add/check if domain exists in the field
+      const used_on = (currentUsedOnValue === null || currentUsedOnValue === "" ||  currentUsedOnValue === undefined)
+        ? `${currentPageHostName},` : relayContextMenus.utils.addUsedOnDomain(currentUsedOnValue, currentPageHostName);
```
Comment (Vinnl) [inline src/js/background/context-menu.js]: Maybe the logic described in the comment could be moved into `addUsedOnDomain`? I think it'll be relevant every time it gets called, no?

File: src/js/background/context-menu.js [hunk 4]
```diff
@@ -153,15 +172,42 @@ const relayContextMenus = {
...
+        auth:true,
+        mask_type: currentMaskType,
+      };
+
+      // Save what site this mask was used on before filling input
+      if (serverStoragePref) {
+        await patchMaskInfo("PATCH", parseInt(selectedAliasId, 10), data, options);
+      } else {
+        selectedAliasObject[0].used_on = used_on;
+        browser.storage.local.set({ relayAddresses: relayAddresses });
```
Comment (Vinnl) [inline src/js/background/context-menu.js:249]: I might be missing a bit too much context in the PR review view, but `relayAddresses` doesn't seem to get changed? Are you sure there's a need here to save it to local storage again?

File: src/js/background/context-menu.js [hunk 5]
```diff
@@ -300,10 +355,15 @@ const relayContextMenus = {
...
-      return relayAddresses.some(alias => website === alias.generated_for);
+    checkIfAnyMasksWereGeneratedOrUsedOnCurrentWebsite: async (website) => {
+      const relayAddresses = await relayContextMenus.utils.getAliases();
+
+      // Short circuit if not current site detected
+      if (website === null) {
+        return false;
+      }
+
+      return relayAddresses.some(alias => website === alias.generated_for || alias.used_on?.includes(website));
```
Comment (Vinnl) [inline src/js/background/context-menu.js]: Probably an edge case, but wouldn't this potentially match e.g. a mask used on `github.io`, even though the current user is at `othersite.github.io`? We miight want to do exact matches:

```suggestion
      return relayAddresses.some(alias => website === alias.generated_for || alias.used_on?.split(",").includes(website));
```

(Likewise for the other instances of `includes` below.)

**Review by Vinnl — APPROVED** [review]
Not sure if I was supposed to review this as well, but since it was assigned to me... :) Just added a question about potential erroneous partial matches.

Thanks for splitting out the formatting changes into a separate commit!

File: src/js/background/context-menu.js (hunk 4 above)
Comment (maxxcrawford) [reply src/js/background/context-menu.js:249]: The `selectedAliasObject[0].used_on` is modified, which is part of `relayAddresses` object. We're making sure to save the new `used_on` entry for the selected/used mask. 

I confirmed this by outputting `relayAddresses` before and after the `used_on` setting on L207.

Comment (maxxcrawford) [reply src/js/background/context-menu.js:249]: I'll add a comment to make this more visible!

File: src/js/background/context-menu.js (hunk 3 above)
Comment (maxxcrawford) [reply src/js/background/context-menu.js]: Ahhh. Great callout. Migrated logic (and renamed the function to `checkAndStoreUsedOnDomain()`) in both the context menu and the in-page menu scripts.

Revised in cf7b8922020cc1f389f341b503d3362b4cfa4132

File: src/js/background/context-menu.js (hunk 5 above)
Comment (maxxcrawford) [reply src/js/background/context-menu.js]: Per conversation on Slack with Tony: 

> Hey reviewing right now. and i think that's okay, can we try releasing it like this and if we get feedback that it's messing users up or causing confusion we can file a bug to change it?

Adding a comment to capture this bug for future us. Comment added in 146b305533e54ea8a749cab984c3f61f891ab852

Comment (maxxcrawford) [reply src/js/background/context-menu.js]:
```
  // BUG: Using includes without spliting on ",", there can 
  // be some false positive results for entries that share the same root domain. 
  // eg: Github mask would show up on example.github.io
  // See full conversation: https://github.com/mozilla/fx-private-relay-add-on/pull/342#discussion_r897755698
```

**Pull request merged**